Fix marriage divorce to allow in cases where user ids are not matched…

Kengetsu opened the pull request:
Fix marriage divorce to allow in cases where user ids are not matched across marriages. This will clear any marriage in a chain where invoking user spouse doesn't match and spouse matches another. All 3 users will have spouses cleared.

levimeahan commented on pages/marriage.php:
```diff
@@ -148,10 +148,49 @@ function marriage() {
+            $result = $system->query("SELECT user_id, user_name, spouse FROM users WHERE user_id ='$player->spouse' LIMIT 1");
+            if (!$system->db_last_num_rows)
+            {
+                throw new Exception("Failed to fetch marriage information");
```
I think in this case we should clear the player's spouse, otherwise we won't solve the one-sided marriage when the other person isn't married

levimeahan commented on pages/marriage.php:
```diff
@@ -148,10 +148,49 @@ function marriage() {
+                    throw new Exception("Error processing divorce!");
+                }
             }
+            else
+           {
+               $spouses_spouse_result = $system->query("SELECT user_id, user_name, spouse FROM users WHERE user_id =" . $current_marriage['spouse'] . " LIMIT 1");
+               if (!$system->db_last_num_rows)
+               {
+                   throw new Exception("Failed to fetch spouse information");
+               }
+               $spouses_spouse = $system->db_fetch($spouses_spouse_result);
+               // Players marriage is invalid and spouse has invalid marriage
+               if (!$current_marriage['user_id'] === $spouses_spouse['spouse'])
+               {
+                   $system->query("UPDATE `users` SET `spouse`='0', `marriage_time`='0' WHERE `user_id`='$player->spouse' OR `user_id`=" . $current_marriage['spouse'] . " LIMIT 2");
+                   if(!$system->db_last_affected_rows) {
+                       throw new Exception("Error processing divorce!");
+                   }
+               }
+               $spouse_player_result = $system->query("SELECT user_id, user_name, spouse FROM users WHERE spouse ='$player->user_id' LIMIT 1");
+
+               // Another player has an unexpected marriage with player
+               if ($system->db_last_num_rows)
+               {
+                   $spouse_player = $system->db_fetch($spouse_player_result);
+
+                   $system->query("UPDATE `users` SET `spouse`='0', `marriage_time`='0' WHERE `user_id`=". $spouse_player['user_id'] ." LIMIT 1");
+                   if(!$system->db_last_affected_rows) {
+                       throw new Exception("Error processing divorce!");
+                   }
+               }
+           }
```
I think we can drop this whole logic, if the spouse is not marked as married to the current player we can just clear the current player's spouse field, and the other person can sort it out on their end

Kengetsu commented:
Reverted all code changes and just removed the throw exception. Now it will just default to removing the player spouse regardless of it updating the other person or not. 

Personally I think keeping the data correct in the database is the better approach but I can definitely see that it adds additional complexity.

levimeahan commented:
Ah sorry for not clarifying - I only meant the logic in the `else` block, but I like the change to check for if player's spouse is also married to them before clearing. We can have a couple scenarios:

Case 1 - Normal marriage
- A married to B
- B married to A
Action: clear both

Case 2 - Love triangle
- A married to B
- B married to C
- C married to B
Action: Clear A

Case 3 - Orphan marriage
- A married to B
- B not married to anyone
Action: Clear A

I will put a comment in with the version of your code I meant

Kengetsu commented on pages/marriage.php:
```diff
@@ -149,9 +149,9 @@ function marriage() {
     if(isset($_POST['confirm_divorce'])) {
         try {
             $system->query("UPDATE `users` SET `spouse`='0', `marriage_time`='0' WHERE `user_id`='$player->spouse' LIMIT 1");
```
Added new commit for this. This will work only change I made is to intval the spouse info so the type comparison works. The only thing I will note from my testing is this won't clear any 3 way mixed up marriages. A -> B, B -> C, C -> A. In those cases it will only clear out the person who clicked divorce. The other two people will still be married on their side but will need to do the divorce as well. I left the other code commented out in case we want to bring it back in. That will check all cases and clear everything that is invalid.

levimeahan replied:
That makes sense. I'd be open to that but we have to make sure cases like the orphaned marriage are fixable (see other comment)

Kengetsu replied:
Makes sense. I can add some more comments into that other code to see if I can get it to make sense or if you have any other input for it. The code that I had is designed to check the player to determine if the marriage is correct in multiple cases.

1. Check the spouse of the player to determine who they are married to. If married to the player it's a correct marriage and divorce is handled as intended. 
2. If the spouse is married to someone else, we check the spouse's spouse (this would be the second chain of marriage) if that matches we ignore that marriage.
3. If that marriage is also invalid, we clear that chain of marriages. (player's spouse and the user married to them).
4. We do a final check to see if anyone is married to the player, if so we clear that as well.

This will clear any 3 link marriages (A -> B, B -> C, C -> A) it won't handle anything more messed up than that but I did pretty extensive testing before in my dev environment to ensure it won't clear valid marriages between players.
